### ISSUE TITLE
Cannot install APK using installApp() method

### ISSUE THREAD

**Issue by deadmoto** [appium/appium#6695]
## The problem

I am trying to install an APK file on an Android device, but I am always getting the same error:

```
org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: packageAndLaunchActivityFromManifest failed. Original error: Path must be a string. Received null (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 50 milliseconds
Build info: version: '2.53.0', revision: '35ae25b1534ae328c771e0856c93e187490ca824', time: '2016-03-15 10:43:46'
System info: host: 'SFOC02RR0UFG8WN.sea.corp.expecn.com', ip: '172.17.18.190', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.6', java.version: '1.8.0_102'
Driver info: io.appium.java_client.android.AndroidDriver
Capabilities [{appPackage=com.hotwire.hotels.debug, networkConnectionEnabled=true, noSign=true, warnings={}, databaseEnabled=false, deviceName=192.168.56.101:5555, platform=LINUX, deviceUDID=192.168.56.101:5555, appActivity=com.hotwire.hotels.home.view.HomePageActivity, desired={appPackage=com.hotwire.hotels.debug, appActivity=com.hotwire.hotels.home.view.HomePageActivity, androidCoverage=com.hotwire.hotels.debug.test/com.hotwire.hotels.debug.test.HotwireInstrumentation, noSign=true, platformName=Android, deviceName=Android Emulator}, androidCoverage=com.hotwire.hotels.debug.test/com.hotwire.hotels.debug.test.HotwireInstrumentation, platformVersion=4.4.4, webStorageEnabled=false, locationContextEnabled=false, takesScreenshot=true, javascriptEnabled=true, platformName=Android}]Session ID: bf65c670-b7ea-42ec-94c3-2b60dcea4ca5
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
    at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
    at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
    at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
    at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:51)
    at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
    at io.appium.java_client.android.AndroidDriver.execute(AndroidDriver.java:1)
    at io.appium.java_client.AppiumDriver.installApp(AppiumDriver.java:295)
```
## Environment
- Appium version (or git revision) that exhibits the issue: v1.5.3 (REV 2b9a2833bc3c6bc8e7b7344e8178ce98e29ebe4b)
- Last Appium version that did not exhibit the issue (if applicable): N/A
- Desktop OS/version used to run Appium: Mac OS X 10.11.6
- Node.js version (unless using Appium.app|exe): v6.3.0
- Mobile platform/version under test: Android 6.0.1 / Android 4.4.4
- Real device or emulator/simulator: Nexus 5X / Genymotion Emulator
- Appium CLI or Appium.app|exe: Appium CLI
## Details

I am trying to install an APK file using following snippet (path modified):

```
androidDriver.installApp("/Users/qa/Downloads/com.android.chrome-1.apk");
```
## Link to Appium logs

https://gist.github.com/deadmoto/cd747718c6da37c3b3d0455bf58f438f

**Comment by qaDream77** [comment]
Rename your apk to something that doesn't have symbols or spaces. Only numbers and letters and ends in .apk.

Also post your capabilities

**Comment by deadmoto** [comment]
@qaDream77 I changed path to APK file to `/Users/Shared/chrome.apk` but it did not help.
My DesiredCapabilities are defined as following (not an actual code):
platformName -> Android
platformVersion -> 4.4.2
deviceName -> emulator-5554
app -> /Users/v-dberezhnoy/Downloads/android-native-app/main/HotwireHotels/build/outputs/apk/HotwireHotels-debug.apk
appPackage -> com.hotwire.hotels.debug
appActivity -> com.hotwire.hotels.home.view.HomePageActivity
androidCoverage -> com.hotwire.hotels.debug.test/com.hotwire.hotels.debug.test.HotwireInstrumentation
noSign -> true

**Comment by mykola-mokhnach** [comment]
@deadmoto  Make sure you get correct output if you launch aapt tool for your package. Appium expects some regexps to have matches after you execute the following command:
```bash
appt dump badging <path_to_apk>
```
These regexps are:
```javascript
new RegExp(/package: name='([^']+)'/g)
```
and
```javascript
RegExp(/launchable-activity: name='([^']+)'/g)
```

In case this output does not contact the necessary information then Appium tries to execute
```bash
java -jar appium_apk_tools.jar printLaunchActivity <apk_path>
```
command. If this command fails as well then the error above is raised. See the 
_manifestMethods.packageAndLaunchActivityFromManifest_ method implementation inside _android-manifest.js_ script for more details.

**Comment by xiaozi0lei** [comment]
hi, guys, I meet the same error.

I use this java code
```
String project_dir = System.getProperty("user.dir");
driver.installApp( project_dir + "/apps/DKSingleSDK_DemoProject_portrait.apk" );
```

log is here

```java
org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: path must be a string (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 39 milliseconds
Build info: version: '2.53.0', revision: '35ae25b1534ae328c771e0856c93e187490ca824', time: '2016-03-15 10:43:46'
System info: host: 'guolei-Lenovo', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-36-generic', java.version: '1.8.0_101'
Driver info: io.appium.java_client.android.AndroidDriver
Capabilities [{app=/home/guolei/gitlab/djsdk/apps/DKSingleSDK_DemoProject_portrait.apk, networkConnectionEnabled=true, warnings={}, apppackage=DKSingleSDK_DemoProject_portrait.apk, databaseEnabled=false, deviceName=QMS4C16114012037, version=5.1.1, platform=LINUX, deviceUDID=QMS4C16114012037, browser_name=, desired={app=/home/guolei/gitlab/djsdk/apps/DKSingleSDK_DemoProject_portrait.apk, browser_name=, appactivity=, autoLaunch=false, apppackage=DKSingleSDK_DemoProject_portrait.apk, unicodeKeyboard=true, platformName=Android, deviceName=Android, version=5.1.1, resetKeyboard=true}, appactivity=, autoLaunch=false, platformVersion=5.1.1, webStorageEnabled=false, locationContextEnabled=false, takesScreenshot=true, javascriptEnabled=true, unicodeKeyboard=true, platformName=Android, resetKeyboard=true}]
Session ID: 67b3a0a5-82de-40be-a1c9-511fa5b0f0dd
```

Anyone can help? Thank you so much!

**Comment by xiaozi0lei** [comment]
My log details is here
```java
[HTTP] --> POST /wd/hub/session/f1b2f293-8d92-4033-99c0-96ac59f33aff/appium/device/install_app {"appPath":"/home/guolei/gitlab/djsdk/apps/DKSingleSDK_DemoProject_portrait.apk"}
[MJSONWP] Calling AppiumDriver.installApp() with args: ["/home/guolei/gitlab/djsdk...
[ADB] Checking whether aapt is present
[ADB] Using aapt from /home/guolei/Android/Sdk/build-tools/23.0.2/aapt
[ADB] Extracting package and launch activity from manifest
[debug] [ADB] badging package: com.duoku.platform.demo.single
[debug] [ADB] badging act: cn.cmgame.billing.api.GameOpenActivity
[MJSONWP] Encountered internal error running command: TypeError: path must be a string
    at TypeError (native)
    at Object.fs.open (fs.js:575:11)
    at ReadStream.open (fs.js:1752:6)
    at new ReadStream (fs.js:1739:10)
    at Object.fs.createReadStream (fs.js:1687:10)
    at module.exports (/home/work/github/appium/node_modules/md5-file/index.js:33:18)
    at tryCatcher (/home/work/github/appium/node_modules/bluebird/js/main/util.js:26:23)
    at ret (eval at <anonymous> (/home/work/github/appium/node_modules/bluebird/js/main/promisify.js:1:0), <anonymous>:13:39)
    at Object.md5$ (lib/fs.js:36:18)
    at tryCatch (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at invoke (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at enqueueResult (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:185:17)
    at new Promise (/home/work/github/appium/node_modules/core-js/library/modules/es6.promise.js:201:7)
    at AsyncIterator.enqueue (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:184:12)
    at AsyncIterator.prototype.(anonymous function) [as next] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at Object.runtime.async (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:209:12)
    at Object.md5 (lib/fs.js:35:32)
    at Object.callee$0$0$ (lib/android-helpers.js:258:25)
    at tryCatch (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at invoke (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at enqueueResult (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:185:17)
    at new Promise (/home/work/github/appium/node_modules/core-js/library/modules/es6.promise.js:201:7)
    at AsyncIterator.enqueue (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:184:12)
    at AsyncIterator.prototype.(anonymous function) [as next] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at Object.runtime.async (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:209:12)
    at Object.callee$0$0 [as installApkRemotely] (lib/android-helpers.js:262:16)
    at AndroidDriver.callee$0$0$ (lib/commands/general.js:105:25)
    at tryCatch (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/home/work/github/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at run (/home/work/github/appium/node_modules/core-js/library/modules/es6.promise.js:108:47)
    at /home/work/github/appium/node_modules/core-js/library/modules/es6.promise.js:119:28
    at flush (/home/work/github/appium/node_modules/core-js/library/modules/$.microtask.js:19:5)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
[HTTP] <-- POST /wd/hub/session/f1b2f293-8d92-4033-99c0-96ac59f33aff/appium/device/install_app 500 490 ms - 192
```

**Comment by johnsba1** [comment]
I have a error more similar to xiaozi0lei where it seems to crash on an md5 hash of the app file. My stack trace looks nearly identical. Here's my setup:

- Appium version : 1.6.0
- Last Appium version that did not exhibit the issue (if applicable): Not sure.
- Desktop OS/version used to run Appium: Mac OS X 10.11.6
- Node.js version (unless using Appium.app|exe): v5.2.0
- Mobile platform/version under test: Android 6.0.1 
- Real device or emulator/simulator: Genymotion Emulator
- Appium CLI or Appium.app|exe: Appium CLI

Stack trace:
[https://gist.github.com/johnsba1/2ee9124fb0fb72e1516aa915db6bfc8e](https://gist.github.com/johnsba1/2ee9124fb0fb72e1516aa915db6bfc8e)


@mykola-mokhnach I went through dumping the badging for my app and I can verify that **package-name** and **launchable-activity** are set. I played with placing the app in /tmp/MyApp.apk to removing spacing/special character issues with no success.

I took a little bit of time to look through the stack, and I didn't get very far before I saw this:

**appium-android-driver:general.js**
```
commands.installApp = async function (appPath) {
  if (!(await fs.exists(appPath))) {
    log.errorAndThrow('Could not find app apk at ${appPath}');
  }

  let {apkPackage} = await this.adb.packageAndLaunchActivityFromManifest(appPath);
  let opts = {
    localApkPath: appPath,
    pkg: apkPackage,
    fastReset: this.opts.fastReset
  };
  return androidHelpers.installApkRemotely(this.adb, opts);
};
```
Then calls **installApkRemotely**:

**appium-android-driver:android-helpers.js**
```
helpers.installApkRemotely = async function (adb, opts) {
  let {app, appPackage, fastReset, androidInstallTimeout} = opts;

  let apkMd5 = await fs.md5(app);
  let remotePath = await helpers.getRemoteApkPath(apkMd5, app);
  let remoteApkExists = await adb.fileExists(remotePath);
  logger.debug("Checking if app is installed");
...
```

It looks like there's a naming mismatch going on in the _opts_ variable. If I change _localApkPath_ to _app_ (and also change _pkg_ to _appPackage_) in general.js then my app tests run and pass.

But then again, I'm no expert on the codebases, so I may be missing something obvious here.
